onyo/commands/new.py

```diff
@@ -1,154 +1,296 @@
+        sys.exit(1)
+    if repo._is_protected_path(asset):
+        print(f"The path is protected by onyo: '{asset}'", file=sys.stderr)
+        sys.exit(1)
 
 
-def run_onyo_new(directory, template, interactive, repo: Repo):
+def create_assets_in_destination(assets: dict, repo: Repo) -> None:
     """
-    Read a new asset name, check it for uniqueness and validity, create the
-    asset (from a template if one was given), edit the new asset (if not
-    suppressed with flag) and stage it with git.
-
-    Returns the newly created asset's filename on success.
+    Create assets and set their contents, and if their directory does not exist
+    yet, create it.
     """
+    for asset in assets.keys():
+        # create missing directories
+        if not asset.parent.is_dir():
+            repo.mkdir(asset.parent)
+        asset.touch()
+        # TODO: this call must be updated when set.py is refactored.
+        set_value(str(asset), str(asset), assets[asset], repo.root)
+        repo.add(asset)
 
-    new_asset = Path(directory, create_filename(repo))
-    temp_asset = Path(repo.root, ".onyo/temp/", new_asset.name)
 
-    template = Path(template).resolve()
-    if template.is_file():
-        shutil.copyfile(template, temp_asset)
-    else:
-        log.error(f"'{template}' does not exist.")
-        sys.exit(1)
+def read_assets_from_tsv(tsv: str, template: Path, set_values: dict, repo: Repo) -> dict:
+    """
+    Read a tsv table with a header row and one row for each new asset to
+    create. Check the information (e.g. filenames correct and unique), and add
+    faux serial numbers to the name if needed.
 
-    if new_asset.exists():
-        log.error(f"'{new_asset}' asset already exists.")
-        sys.exit(1)
+    Returns a dictionary with paths and contents of assets on success, or
+    informs the user about missing or invalid information and exits with error.
+    """
+    new_assets = {}
+    existing_assets = repo.assets
+    row_count = 0
+
+    with open(tsv, 'r') as tsv_file:
+        # count and request the faux serial numbers needed
+        table = csv.DictReader(tsv_file, delimiter='\t')
+        faux_number = sum([1 for row in table if row['serial'] == 'faux'])
+        if faux_number > 0:
+            faux_serial_list = repo.generate_faux_serials(num=faux_number)
+
+        # iterate over the table from the beginning to read asset information.
+        tsv_file.seek(0)
+        table = csv.DictReader(tsv_file, delimiter='\t')
+        for row in table:
+            row_count += 1
+            new_path = ""
+            contents = {}
+
+            # error if any required field is empty
+            if not row['type'] or not row['make'] or not row['model'] or not row['serial'] or not row['directory']:
+                print("The fields 'type', 'make', 'model', 'serial' and 'directory' are required, but missing in line {row_count}.", file=sys.stderr)
```

Comment:
> but missing in line

This is an incomplete clause.

From a user standpoint, I think it would be helpful to also print the entire offending line.

Reply:
To shame me even more, it also should have better been an f-string.

I added the full row, thanks.